File: 2025/day05-1.hs
import Data.List (find)
import Data.Maybe (fromJust, isJust)
import Debug.Trace (traceShow)

type Range = (Int, Int)

type Id = Int

solution :: String -> Int
solution content = length . filter (isFresh ranges) $ ids
  where
    ranges = reduce . map parseRange . takeWhile (/= "") . lines $ content
    ids = map read . tail . dropWhile (/= "") . lines $ content

isFresh :: [Range] -> Id -> Bool
isFresh rs id = any (isWithin id) rs

isWithin :: Id -> Range -> Bool
isWithin id (start, end) = start <= id && id <= end

reduce :: [Range] -> [Range]
reduce [] = []
reduce (r : rs)
  | isJust other = reduce (merge r justOther : (filter (/= justOther) . filter (/= r) $ rs))
  | otherwise = r : reduce rs
  where
    other = find (isMergable r) rs
    justOther = fromJust other

isMergable :: Range -> Range -> Bool
isMergable a b = helper a b || helper b a
  where
    helper (start1, end1) (start2, end2) = start2 <= start1 && start1 <= end2 || start1 <= start2 && end2 <= end1

merge :: Range -> Range -> Range
merge (start1, end1) (start2, end2) = (min start1 start2, max end1 end2)

parseRange :: String -> Range
parseRange line = (read . takeWhile (/= '-') $ line, read . tail . dropWhile (/= '-') $ line)

main :: IO ()
main = do readFile "day05.dat" >>= print . solution

File: 2025/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Range = tuple[int, int]


def solution(content: str) -> Any:
    range_block, ingredient_block = content.split("\n\n")
    ranges = [parse_range(line) for line in range_block.split("\n")]
    ingredients = [int(x) for x in ingredient_block.split("\n")]
    return len(
        [ingredient for ingredient in ingredients if is_fresh(ranges, ingredient)]
    )


def is_fresh(ranges: list[Range], ingredient: int) -> bool:
    return any(start <= ingredient <= end for start, end in ranges)


def parse_range(line: str) -> Range:
    start, _, end = line.partition("-")
    return (int(start), int(end))


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
